### PR TITLE
Modernizes tgui input list

### LINKED ISSUE #63344: TGUI Input lists do not let you use the arrow keys to move up and down
Regression from older lists, important for a11y. 

Pressing up/down should move the cursor up and down, currently it does nothing..

### LINKED ISSUE #63350: Show camera list bluescreen
## Testmerges:

#63331 

## Reproduction:

Typing anything in the "Show Camera list" Ui for AI will quickly result in a blue screen for AIs
image

image

Scrolling up and down the menu works at least, I guess?

### LINKED ISSUE #63400: Ghost Teleport Menu BSOD
Reporting client version: 514.1568

## Round ID:
[174757](https://scrubby.melonmesa.com/round/174757)

## Testmerges:
- [Revert "Require Node 16"](https://www.github.com/tgstation/tgstation/pull/63331)

## Reproduction:
image
Typed something wrong, hit backspace to delete, accidentally hit \\, bluescreen

### PR DESCRIPTION
**Pull request by jlsnow301**
## About The Pull Request
Converts tgui_input_list to modern typescript react.
Now responds to preferences in button sizing and placement.
Search bar default visibility is based on list size (idea courtesy @GoldenAlpharex)

5xszlt

## Why It's Good For The Game
Fixes #63344 
Fixes #63350
#63390 (fixes this but i want the other PR to close it)
Fixes #63400
List input now obeys having large / swapped buttons in preferences
Better accessibility as down/up/enter keys are working
I don't think this was implemented before either: You can now press a letter to scroll to it (if it exists as a list item)

## Changelog

:cl:
fix: TGUI list input will now respond to having large/swapped buttons in preferences.
fix: BYOND functionality restored in TGUI input lists - keyboard support. Up/Down, type a letter to jump to it (while search is not visible).
/:cl:

### CODE REVIEW

**Comment by GoldenAlpharex** [comment]
Very cool! Just making sure, when the list is down, using keys will jump to the first item that starts with that key, right? I don't remember if that was functional or not, but that's also default BYOND input list behavior.

**Comment by jlsnow301** [comment]
> Very cool! Just making sure, when the list is down, using keys will jump to the first item that starts with that key, right? I don't remember if that was functional or not, but that's also default BYOND input list behavior.

I can probably code this functionality in, but as of yet, no

**Comment by jlsnow301** [comment]
things i'm not necessarily happy with - 
focus() sort of lags behind so it will clip the top and bottom entries if you loop through the list (ie, press up on index 0 => index len) 
i don't think i can disable hotkeys for character movement while using the screen so hotkeying to an entry (ex: press a) will still make your character move

File: tgui/packages/tgui/interfaces/ListInputModal.tsx [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+  const [searchQuery, setSearchQuery] = useSharedState<string>(
+    context,
+    'searchQuery',
+    ''
+  );
+  const [inputIsValid, setInputIsValid] = useSharedState<Validator>(
+    context,
+    'inputIsValid',
+    { isValid: true, error: null }
+  );
```
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/ListInputModal.tsx:39]: None of these should be `useSharedState` (I'm hesitant to say anything should be), use local state.

**Comment by stylemistake** [comment]
I will just say this, that meme is inappropriate, because class is really much more superior from performance standpoint. And if you have lags at any point, it's because of this.

**Comment by jlsnow301** [comment]
> You shan't forget that this is Inferno and not React, which is why we can't have nice things, but if you do things right, it can be much faster than React.

Interesting. I am indeed looking at this from a React angle. I prefer SFCs almost entirely on readability (but that readability depends on where you learned it, I suppose)

Personally I find the this keyword a little opaque at times and don't use redux at all in any of my personal projects (vs useContext) so I wasn't considering its performance thereof

**Comment by Tastyfish** [comment]
Alrighty, if that fixes the AI camera blue-screen, I imagine the code allows for `null` entries in the items/buttons list now, so my PR https://github.com/tgstation/tgstation/pull/63394 will be redundant. Just making sure this is the case.

**Comment by jlsnow301** [comment]
> Alrighty, if that fixes the AI camera blue-screen, I imagine the code allows for `null` entries in the items/buttons list now, so my PR #63394 will be redundant. Just making sure this is the case.

Sorry I did not see this PR yet. I think it should be left open so that it's taken care of on both sides. This simply skips over null entries

**Review by stylemistake — APPROVED** [review]

**Pull request merged**